scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -9,22 +9,32 @@
             [cook.util :as util]
             [clojure.string :as str]
             [clojure.tools.logging :as log]
-            [metrics.counters :as counters]
             [metrics.timers :as timers])
   (:import (clojure.lang IAtom)
            (io.kubernetes.client.models V1Pod V1ContainerStatus V1PodStatus)
            (java.net URLEncoder)))
 
-;
-;   Wire up a store with the results.
-;
-
-(def lock-object (Object.))
-(defn calculate-lock
-  "Given a pod-name, return an object suitable for locking for accessing it."
-  [_]
-  ; TODO: Should do lock sharding based on hash of pod-name.
-  lock-object)
+(defmacro with-process-lock
+  "Evaluates body (which should contain a call to
+  process) after acquiring the pod-name-based lock"
+  [compute-cluster-name pod-name & body]
+  `(timers/time!
+     (metrics/timer "process-lock" ~compute-cluster-name)
+     (let [lock-objects#
+           (:controller-lock-objects (config/kubernetes))
+           lock-object#
+           (nth lock-objects#
+                (-> ~pod-name
+                    hash
+                    (mod (count lock-objects#))))
+           timer-context#
+           (timers/start
+             (metrics/timer
+               "process-lock-acquire"
+               ~compute-cluster-name))]
+       (locking lock-object#
```

Comment:
If this is going to be a macro, switch to https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantLock.html do the try/finally. Locking time should be microseconds on an empty queue.

Reply:
I don't understand the "If this is going to be a macro" part of your comment -- isn't the decision to use `ReentrantLock` orthogonal to whether or not this is a macro?

Reply:
Yes and no. Part of the convenience of using a monitor-style lock is conciser syntax and harder to misuse (because you're forced syntatically to free the lock at the end.) However, those advantages are eliminated when using a macro. ReentrantLock is faster, the modern choice,  and has more metrics --- in particular, queue size.

Reply:
Sounds good, will do

Reply:
Done